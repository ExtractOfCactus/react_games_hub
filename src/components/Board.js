import React from 'react';

const COLUMNS = ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H']

function Square(props) {
  return (
    <button
      className="square"
      onClick={props.onClick}
    >
      {props.value}
    </button>
  );
}

function ColumnNotation(props) {
  return (
    <span className="column-notation">
      {props.value}
    </span>
  );
}

function RowNotation(props) {
  return (
    <span className="row-notation">
      {props.value}
    </span>
  );
}

class Board extends React.Component {
//  const size = this.props.size;

  renderSquare(i) {
    return (
      <Square
        value={this.props.squares[i]}
        onClick={() => this.props.onClick(i)}
      />
    );
  }

  renderColumnNotation(i) {
    return (
      <ColumnNotation value={i} />
    );
  }

  renderRowNotation(i) {
    return (
      <RowNotation value={i} />
    );
  }

  renderBoardColumns() {
    let columnHeaders = [];
    for (let i = 0; i < this.props.size; i++) {
      columnHeaders.push(
        this.renderColumnNotation(COLUMNS[i])
      )
    }
    return <div className="board-columns">{columnHeaders}</div>
  }

  renderBoardSquares() {
    let boardSquares = [];
    let rows = [];
    for (let row = 0; row < this.props.size; row++) {
      for (let i = row * this.props.size; i < row * this.props.size + this.props.size; i++) {
        console.log(i)
        boardSquares.push(this.renderSquare(i));
      }
      rows.push(
        <div key={row} className="board-row">
          {this.renderRowNotation(row + 1)}
          {boardSquares}
        </div>
      );
      boardSquares = [];
    }
    return rows;
  }

  render() {
    return (
      <div>
        <div>
          {this.renderBoardColumns()}
          {this.renderBoardSquares()}
        </div>

      </div>
    );
  }
}

export default Board;