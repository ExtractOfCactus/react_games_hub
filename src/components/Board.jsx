import React from 'react';
import Square from './Square'

const COLUMNS = ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H']

function ColumnNotation(props) {
  return (
    <span className="column-notation">
      {props.value}
    </span>
  );
}

function RowNotation(props) {
  return (
    <span className="row-notation">
      {props.value}
    </span>
  );
}

class Board extends React.Component {

  renderSquare(i, className) {
    return (
      <Square
        key={i}
        id={i}
        classOption={className}
        value={this.props.squares[i]}
        onClick={() => this.props.onClick(i)}
      />
    );
  }

  renderColumnNotation(i) {
    return (
      <ColumnNotation key={i} value={i} />
    );
  }

  renderRowNotation(i) {
    return (
      <RowNotation value={i} />
    );
  }

  renderBoardColumns() {
    let columnHeaders = [];
    for (let i = 0; i < this.props.size; i++) {
      columnHeaders.push(
        this.renderColumnNotation(COLUMNS[i])
      )
    }
    return <div className="board-columns">{columnHeaders}</div>
  }

  determineSquareClass(row, index) {
    if (this.props.game === 1) {
      return 'square'
    }
    return row % 2 === 0 ? 
      (index % 2 === 0 ? 'dark-square' : 'light-square'):
      (index % 2 === 0 ? 'light-square' : 'dark-square');
  }

  renderBoardSquares() {
    let boardSquares = [];
    let rows = [];
    for (let row = 0; row < this.props.size; row++) {
      for (let i = row * this.props.size; i < row * this.props.size + this.props.size; i++) {
        const classOption = this.determineSquareClass(row, i)
        boardSquares.push(this.renderSquare(i, classOption));
      }
      rows.push(
        <div key={row} className="board-row">
          {this.renderRowNotation(row + 1)}
          {boardSquares}
        </div>
      );
      boardSquares = [];
    }
    return rows;
  }

  render() {
    return (
      <div>
        <div>
          {this.renderBoardColumns()}
          {this.renderBoardSquares()}
        </div>

      </div>
    );
  }
}

export default Board;
