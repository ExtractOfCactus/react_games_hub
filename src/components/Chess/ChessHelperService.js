const mapPieces = {
  0: 'wR',
  1: 'wN',
  2: 'wB',
  3: 'wQ',
  4: 'wK',
  5: 'wB',
  6: 'wN',
  7: 'wR',
  56: 'bR',
  57: 'bN',
  58: 'bB',
  59: 'bQ',
  60: 'bK',
  61: 'bB',
  62: 'bN',
  63: 'bR',
}

export const setupChessStart = () => {
  let squares = Array(64).fill(null);
  for (let i=0; i < 64; i++) {
    if (i > 7 && i < 16) {
      squares[i] = 'wp';
    } else if (i > 47 && i < 56) {
      squares[i] = 'bp';
    } else if (i < 8 || i > 55) {
      squares[i] = mapPieces[i];
    }
  }
  return squares;
}