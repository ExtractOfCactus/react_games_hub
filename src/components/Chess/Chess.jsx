import React from "react";
import Board from "../Board";
import { setupChessStart } from "./ChessHelperService";
import white_pawn from "../../images/white_pawn.png";

class Chess extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      history: [
        {
          squares: setupChessStart(),
          movePosition: null
        }
      ],
      nextPlayer: "w",
      stepNumber: 0,
      previousFocus: null,
      focus: null
    };
  }

  handleClick(i) {
    let history = this.state.history;
    const current = history[history.length - 1];
    let squares = current.squares.slice()
    let previousFocus = this.state.previousFocus;
    if (previousFocus === null) {
      previousFocus = i;
    } else {
      squares[i] = 'wp';
      squares[previousFocus] = null;
      previousFocus = null;
    }
    this.setState({
      history: [{squares: squares}],
      previousFocus: previousFocus,
    });
  }

  render() {
    const current = this.state.history[this.state.stepNumber];
    const status = "Next player";
    const statusImage = (
      <img alt="White piece" src={white_pawn} className="status-image" />
    );
    return (
      <div>
        <div className="game-board">
          <Board
            squares={current.squares}
            onClick={i => this.handleClick(i)}
            size={8}
          />
        </div>
        <div className="game-info">
          <div>{status}</div>
          <div>{statusImage}</div>
        </div>
        <div>
          <button
            className="game-selector"
            onClick={() => {
              this.props.returnHome(0);
            }}
          >
            Home Page
          </button>
        </div>
      </div>
    );
  }
}

export default Chess;
