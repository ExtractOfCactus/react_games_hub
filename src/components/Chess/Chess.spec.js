import React from "react";
import Chess from "./Chess";
import Board from "../Board";
import { shallow, mount } from "enzyme";
import { setupChessStart } from "./ChessHelperService";

describe("Chess game at start", () => {
  const wrapper = shallow(<Chess />);

  it("renders anything", () => {
    expect(wrapper.exists()).toBe(true);
  });

  it("renders the board", () => {
    expect(wrapper.containsMatchingElement(<Board />)).toEqual(true);
  });

  it("renders the home page button", () => {
    expect(wrapper.containsMatchingElement(<button>Home Page</button>)).toEqual(
      true
    );
  });

  it("renders a div for the status", () => {
    expect(wrapper.find(".game-info").text()).toEqual("Next player");
  });

  it("can populate board with pieces", () => {
    const squares = setupChessStart();
    expect(squares[0]).toEqual("bR");
    expect(squares[4]).toEqual("bK");
    expect(squares[6]).toEqual("bN");
    expect(squares[10]).toEqual("bp");
    expect(squares[30]).toEqual(null);
    expect(squares[52]).toEqual("wp");
    expect(squares[59]).toEqual("wQ");
    expect(squares[61]).toEqual("wB");
    expect(squares[63]).toEqual("wR");
  });

  it("matches snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });
});

describe("A pawn can", () => {
  const wrapper = mount(<Chess />);
  // console.log(wrapper.debug());

  it("be found starting at A2 (square 48)", () => {
    const squareA2 = wrapper.find('#square-48');
    expect(squareA2.find('#wp')).toHaveLength(1);
  });

  it("not be found starting at A3 (square 40)", () => {
    const squareA3 = wrapper.find('#square-40');
    expect(squareA3.find('#wp')).toHaveLength(0);
  });

  it("move forward one space", () => {
      let squareA2 = wrapper.find('#square-48');
      squareA2.simulate('click');
      
      let squareA3 = wrapper.find('#square-40');
      squareA3.simulate('click');

      squareA2 = wrapper.find('#square-48');
      squareA3 = wrapper.find('#square-40');
      expect(squareA2.find('#wp')).toHaveLength(0);
      expect(squareA3.find('#wp')).toHaveLength(1);
  });
});
