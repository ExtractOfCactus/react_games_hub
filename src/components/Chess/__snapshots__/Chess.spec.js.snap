// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chess game at start matches snapshot 1`] = `
<div>
  <div
    className="game-board"
  >
    <Board
      onClick={[Function]}
      size={8}
      squares={
        Array [
          "bR",
          "bN",
          "bB",
          "bQ",
          "bK",
          "bB",
          "bN",
          "bR",
          "bp",
          "bp",
          "bp",
          "bp",
          "bp",
          "bp",
          "bp",
          "bp",
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          "wp",
          "wp",
          "wp",
          "wp",
          "wp",
          "wp",
          "wp",
          "wp",
          "wR",
          "wN",
          "wB",
          "wQ",
          "wK",
          "wB",
          "wN",
          "wR",
        ]
      }
    />
  </div>
  <div
    className="game-info"
  >
    <div>
      Next player
    </div>
    <div>
      <img
        alt="White piece"
        className="status-image"
        src={
          Object {
            "process": [Function],
          }
        }
      />
    </div>
  </div>
  <div>
    <button
      className="game-selector"
      onClick={[Function]}
    >
      Home Page
    </button>
  </div>
</div>
`;
